File: frontend/src/services/writing.ts
import api from './apiClient';

export interface ProgressResponse {
  level: string;
  skill_id?: string;
  exercises_completed: number;
  exercises_total?: number;
  stars_earned: number;
  history?: string[];
}

export interface Session {
  id: string;
  level: string;
  skill_id?: string;
  session_state?: string;
  best_score?: number;
  status: string;
  expires_at?: string;
  created_at: string;
}

export interface ExerciseResponse {
  source: string;
  exercise?: Record<string, unknown>;
  message?: string;
}

export interface EvaluateResponse {
  id: string;
  word_count: number;
  feedback?: Record<string, unknown>;
  message?: string;
}

export interface VocabItem {
  id: string;
  word: string;
  definition?: string;
  cefr_level?: string;
  review_count: number;
}

export interface VocabListResponse {
  words: VocabItem[];
  count: number;
}

export interface ModelEssay {
  id: string;
  topic?: string;
  task_type: string;
  content: string;
  word_count?: number;
  band_score?: number;
  category?: string;
  views_count: number;
  saves_count: number;
}

export interface DevilsAdvocateResponse {
  id: string;
  detected_claim?: string;
  counter_point?: string;
  feedback?: string;
}

export interface PeerSubmission {
  id: string;
  essay_content: string;
  prompt?: string;
  task_type?: string;
  word_count?: number;
  status: string;
  created_at: string;
}

export interface PeerReview {
  id: string;
  submission_id: string;
  task_response_score?: number;
  coherence_score?: number;
  lexical_score?: number;
  grammar_score?: number;
  overall_band?: number;
  strengths?: string;
  weaknesses?: string;
  inline_corrections?: string;
}

export interface ReviewQueueResponse {
  ok: boolean;
  overall_band?: number;
}

export const writingService = {
  async getProgress(): Promise<ProgressResponse[]> {
    const response = await api.get<ProgressResponse[]>('/api/writing/progress');
    return response.data || [];
  },

  async saveProgress(data: {
    level: string;
    skill_id?: string;
    exercises_completed: number;
    stars_earned: number;
    history?: string[];
  }): Promise<{ ok: boolean; error?: string }> {
    const response = await api.post<{ ok: boolean }>('/api/writing/progress', data);
    if (response.error) {
      return { ok: false, error: response.error.error };
    }
    return { ok: true };
  },

  async getSessions(level?: string): Promise<Session[]> {
    const query = level ? `?level=${level}` : '';
    const response = await api.get<Session[]>(`/api/writing/sessions${query}`);
    return response.data || [];
  },

  async saveSession(data: {
    id?: string;
    level: string;
    skill_id?: string;
    session_state?: string;
    best_score?: number;
    status?: string;
    expires_at?: string;
  }): Promise<{ ok: boolean; id?: string; error?: string }> {
    const response = await api.post<{ ok: boolean; id: string }>('/api/writing/sessions', data);
    if (response.error) {
      return { ok: false, error: response.error.error };
    }
    return { ok: true, id: response.data?.id };
  },

  async getExercise(level: string, skillId: string): Promise<ExerciseResponse> {
    const response = await api.post<ExerciseResponse>('/api/writing/exercise', { level, skill_id: skillId });
    return response.data || { source: 'error', message: 'Failed to fetch exercise' };
  },

  async evaluateEssay(data: {
    essay: string;
    task_type: string;
    prompt?: string;
    time_spent_seconds?: number;
  }): Promise<EvaluateResponse> {
    const response = await api.post<EvaluateResponse>('/api/writing/evaluate', data);
    return response.data || { id: '', word_count: 0, message: 'Failed to evaluate' };
  },

  async listModelEssays(taskType?: string): Promise<ModelEssay[]> {
    const query = taskType ? `?task_type=${taskType}` : '';
    const response = await api.get<ModelEssay[]>(`/api/writing/model-essays${query}`);
    return response.data || [];
  },

  async getVocabulary(): Promise<VocabListResponse> {
    const response = await api.get<VocabListResponse>('/api/writing/vocabulary');
    return response.data || { words: [], count: 0 };
  },

  async addVocabulary(data: {
    word: string;
    definition?: string;
    cefr_level?: string;
    example_sentence?: string;
    source_essay_id?: string;
  }): Promise<{ ok: boolean; error?: string }> {
    const response = await api.post<{ ok: boolean }>('/api/writing/vocabulary', data);
    if (response.error) {
      return { ok: false, error: response.error.error };
    }
    return { ok: true };
  },

  async devilsAdvocate(data: {
    user_argument: string;
    time_spent_seconds?: number;
  }): Promise<DevilsAdvocateResponse> {
    const response = await api.post<DevilsAdvocateResponse>('/api/writing/devils-advocate', data);
    return response.data || { id: '' };
  },

  async submitEssay(data: {
    essay_content: string;
    prompt?: string;
    task_type: string;
    is_anonymous?: boolean;
  }): Promise<{ ok: boolean; id?: string; error?: string }> {
    const response = await api.post<{ ok: boolean; id: string }>('/api/writing/peer-review/submissions', data);
    if (response.error) {
      return { ok: false, error: response.error.error };
    }
    return { ok: true, id: response.data?.id };
  },

  async getReviewQueue(): Promise<PeerSubmission[]> {
    const response = await api.get<PeerSubmission[]>('/api/writing/peer-review/queue');
    return response.data || [];
  },

  async submitReview(data: {
    submission_id: string;
    task_response_score: number;
    coherence_score: number;
    lexical_score: number;
    grammar_score: number;
    strengths?: string;
    weaknesses?: string;
    suggestions?: string;
    inline_corrections?: string;
    time_spent_seconds?: number;
  }): Promise<ReviewQueueResponse> {
    const response = await api.post<ReviewQueueResponse>('/api/writing/peer-review/reviews', data);
    return response.data || { ok: false };
  },
};

export default writingService;

File: src/models/writing.rs
use serde::{Deserialize, Serialize};
use vil_orm_derive::VilEntity;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "writing_gym_progress")]
pub struct WritingGymProgress {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub level: String,
    pub skill_id: Option<String>,
    pub exercises_completed: i64,
    pub exercises_total: Option<i64>,
    pub stars_earned: i64,
    pub history: Option<String>, // JSON
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "writing_sessions")]
pub struct WritingSession {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub level: String,
    pub skill_id: Option<String>,
    pub session_state: Option<String>, // JSON
    pub best_score: Option<i64>,
    pub status: String,
    pub expires_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "peer_review_submissions")]
pub struct PeerReviewSubmission {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub essay_content: String,
    pub prompt: Option<String>,
    pub task_type: Option<String>,
    pub word_count: Option<i64>,
    pub is_anonymous: i32,
    pub status: String,
    pub claimed_by: Option<String>,
    pub claimed_at: Option<String>,
    pub difficulty_level: Option<String>,
    pub moderation_status: Option<String>,
    pub report_count: Option<i64>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "peer_reviews")]
pub struct PeerReview {
    #[vil_entity(pk)]
    pub id: String,
    pub submission_id: String,
    pub reviewer_id: String,
    pub task_response_score: Option<i64>,
    pub coherence_score: Option<i64>,
    pub lexical_score: Option<i64>,
    pub grammar_score: Option<i64>,
    pub overall_band: Option<f64>,
    pub strengths: Option<String>,
    pub weaknesses: Option<String>,
    pub suggestions: Option<String>,
    pub inline_corrections: Option<String>,
    pub time_spent_seconds: Option<i64>,
    pub helpfulness_rating: Option<i64>,
    #[vil_entity(auto_now_add)]
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "writing_submissions")]
pub struct WritingSubmission {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub task_type: String,
    pub prompt: Option<String>,
    pub reading_passage: Option<String>,
    pub user_essay: String,
    pub word_count: Option<i64>,
    pub ai_score: Option<i64>,
    pub ai_feedback: Option<String>,
    pub breakdown: Option<String>,
    pub time_spent_seconds: Option<i64>,
    #[vil_entity(auto_now_add)]
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "model_essays")]
pub struct ModelEssay {
    #[vil_entity(pk)]
    pub id: String,
    pub topic: Option<String>,
    pub task_type: String,
    pub content: String,
    pub word_count: Option<i64>,
    pub band_score: Option<f64>,
    pub breakdown: Option<String>,
    pub annotations: Option<String>,
    pub highlights: Option<String>,
    pub category: Option<String>,
    pub source: Option<String>,
    pub views_count: i64,
    pub saves_count: i64,
    #[vil_entity(auto_now_add)]
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "collected_vocabulary")]
pub struct CollectedVocabulary {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub word: String,
    pub definition: Option<String>,
    pub cefr_level: Option<String>,
    pub example_sentence: Option<String>,
    pub source_essay_id: Option<String>,
    pub review_count: i64,
    pub next_review_at: Option<String>,
    pub collected_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "devils_advocate_sessions")]
pub struct DevilsAdvocateSession {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub user_argument: Option<String>,
    pub detected_claim: Option<String>,
    pub counter_point: Option<String>,
    pub score: Option<i64>,
    pub feedback: Option<String>,
    pub time_spent_seconds: Option<i64>,
    #[vil_entity(auto_now_add)]
    pub created_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "exercise_pool")]
pub struct ExercisePoolItem {
    #[vil_entity(pk)]
    pub id: String,
    pub level: String,
    pub skill_id: String,
    pub difficulty: Option<String>,
    pub exercise_data: String,
    #[vil_entity(auto_now_add)]
    pub created_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SubmitEssayRequest {
    pub essay_content: String,
    pub prompt: Option<String>,
    pub task_type: String,
    pub is_anonymous: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct SubmitReviewRequest {
    pub submission_id: String,
    pub task_response_score: i64,
    pub coherence_score: i64,
    pub lexical_score: i64,
    pub grammar_score: i64,
    pub strengths: Option<String>,
    pub weaknesses: Option<String>,
    pub suggestions: Option<String>,
    pub inline_corrections: Option<String>,
    pub time_spent_seconds: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct SaveProgressRequest {
    pub level: String,
    pub skill_id: Option<String>,
    pub exercises_completed: i64,
    pub stars_earned: i64,
    pub history: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SessionFilter {
    pub level: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SaveSessionRequest {
    pub id: Option<String>,
    pub level: String,
    pub skill_id: Option<String>,
    pub session_state: Option<String>,
    pub best_score: Option<i64>,
    pub status: Option<String>,
    pub expires_at: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ExerciseRequest {
    pub level: String,
    pub skill_id: String,
}

#[derive(Debug, Deserialize)]
pub struct EvaluateEssayRequest {
    pub essay: String,
    pub task_type: String,
    pub prompt: Option<String>,
    pub time_spent_seconds: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct ModelEssayFilter {
    pub task_type: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct AddVocabRequest {
    pub word: String,
    pub definition: Option<String>,
    pub cefr_level: Option<String>,
    pub example_sentence: Option<String>,
    pub source_essay_id: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct DevilsAdvocateRequest {
    pub user_argument: String,
    pub time_spent_seconds: Option<i64>,
}
